File: suat/0026.py
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        head, end = 0, 0

        while end < len(nums):
            if nums[head] == nums[end]:
                end += 1
            else:
                head += 1
                nums[head], nums[end] = nums[end], nums[head]
                end += 1

        return head +1




'''
R
0,0,1,1,1,2,2,3,3,4
L

    R
0,0,1,1,1,2,2,3,3,4
L

shift L, change numbers, shift R
      R
0,1,0,1,1,2,2,3,3,4
  L

          R
0,1,0,1,1,2,2,3,3,4
  L  

            R
0,1,2,1,1,0,2,3,3,4
    L 


'''

File: suat/0026.dart

main() {
  var result = Solution().removeDuplicates([0, 0, 1, 1, 1, 2, 2, 3, 3, 4]);
  print(result);
}

class Solution {
  /// Using a two pointer approach, head for holding resulting array, end for shifting till the end.
  int removeDuplicates(List<int> nums) {
    int head = 0; // pointer
    int end = 0;

    while (end < nums.length) {
      if (nums[end] > nums[head]) {
        head++;
        nums[head] = nums[end];
      }
      end++;
    }

    return head + 1; // return not pointing index, number of elements
  }
}
